components/modules/eromfs.c

```diff
@@ -0,0 +1,377 @@
+#include "module.h"
+#include "lauxlib.h"
+
+#include "esp_vfs.h"
+#include <errno.h>
+#include <dirent.h>
+
+#include <stdint.h>
+#include <string.h>
+
+/**
+ * The logical layout of the embedded volumes is
+ * [ volume record ]
+ * [ volume record ]
+ * ...
+ * [ index record in vol1 ]
+ * [ index record in vol1 ]
+ * ...
+ * [ file contents in vol1 ]
+ * [ file contents in vol1 ]
+ * ...
+ * [ index record in vol2 ]
+ * [ index record in vol2 ]
+ * ...
+ * [ file contents in vol2 ]
+ * [ file contents in vol2 ]
+ *
+ * Both the volume records and index records are variable length so as to not
+ * waste space in their name fields. Finding the start of the index records
+ * for a volume is by reading the offs(et) field in the volume record and
+ * jumping that many bytes forward from the start of the eromfs.bin data.
+ * Similarly, finding the file contents is by reading the index record's
+ * offs(et) field and basing that off the start of the volume index.
+ * Naturally, the start of the volume index is the same as the end of the
+ * volume header, and the start of the file contents is the same as the
+ * end of the volume index, and either of those can be worked out by
+ * reading the offs(et) field in the first record.
+ */
+
+#pragma pack(push, 1)
+typedef struct {
+  uint8_t  rec_len;
+  uint16_t offs; // index_offs
+  char     name[];
+} volume_record_t;
+
+typedef struct {
+  uint8_t  rec_len;
+  uint32_t offs; // based off index_offs
+  uint32_t len; // file_len
+  char     name[];
+} index_record_t;
+#pragma pack(pop)
+
+
+typedef struct {
+  const index_record_t *meta;
+  const char *data; // start of data
+  off_t pos;
+} file_descriptor_t;
+
+
+typedef struct {
+  DIR opaque;
+  const index_record_t *index;
+  const index_record_t *pos;
+} eromfs_DIR_t;
+
+
+extern const char eromfs_bin_start[] asm("_binary_eromfs_bin_start");
+
+// Both the volume header and the file set indices end where the next
+// type of data block commences (file set index, file contents).
+#define end_of(type, start) ((const type *)(((char *)start) + start->offs))
+
+#define eromfs_header_start ((const volume_record_t *)eromfs_bin_start)
+#define eromfs_header_end   end_of(volume_record_t, eromfs_header_start)
+
+
+/* The logic for finding a volume record by name is the same as finding a
+ * file record by name, only the data structure type varies. Hence we
+ * hide the casting and variable length record stepping behind a convenience
+ * macro here.
+ */
+#define find_entry_by_name(out, xname, start_void_p, record_t) \
+  do { \
+    const record_t *entry_ = (const record_t *)(start_void_p); \
+    const record_t *end_ = end_of(record_t, entry_); \
+    for (; entry_ < end_; \
+         entry_ = (const record_t *)(((char *)entry_) + entry_->rec_len)) \
+    { \
+      uint8_t name_len = entry_->rec_len - sizeof(record_t); \
+      if (strncmp(xname, entry_->name, name_len) == 0) \
```

Comment:
This doesn't quite work -- if xname is "foobar" and name is foo (with namelen 3), you will get a match.

Reply:
Good call-out! There's a length comparison missing there. Will fix.